File: modules/agent/src/lib/storage.py
"""Storage abstraction for document file storage.

Supports:
- LocalStorage: Filesystem storage for local development
- R2Storage: Cloudflare R2 storage for production (S3-compatible)
"""

import os
from abc import ABC, abstractmethod
from pathlib import Path

import aiofiles
import boto3


class StorageBackend(ABC):
    """Abstract base class for storage backends."""

    @abstractmethod
    async def upload(self, path: str, data: bytes, content_type: str) -> str:
        """Upload file, return storage path."""

    @abstractmethod
    async def download(self, path: str) -> bytes:
        """Download file content."""

    @abstractmethod
    async def delete(self, path: str) -> None:
        """Delete file."""

    @abstractmethod
    def get_url(self, path: str, expires_in: int = 3600) -> str:
        """Get download URL (signed for R2, file:// for local)."""

    @abstractmethod
    async def exists(self, path: str) -> bool:
        """Check if file exists."""


class LocalStorage(StorageBackend):
    """Filesystem storage for local development."""

    def __init__(self, base_path: str = "./storage/documents"):
        self.base_path = Path(base_path)
        self.base_path.mkdir(parents=True, exist_ok=True)

    async def upload(self, path: str, data: bytes, content_type: str) -> str:
        full_path = self.base_path / path
        full_path.parent.mkdir(parents=True, exist_ok=True)
        async with aiofiles.open(full_path, "wb") as f:
            await f.write(data)
        return path

    async def download(self, path: str) -> bytes:
        async with aiofiles.open(self.base_path / path, "rb") as f:
            return await f.read()

    async def delete(self, path: str) -> None:
        file_path = self.base_path / path
        if file_path.exists():
            file_path.unlink()

    def get_url(self, path: str, expires_in: int = 3600) -> str:
        return f"file://{(self.base_path / path).absolute()}"

    async def exists(self, path: str) -> bool:
        return (self.base_path / path).exists()


class R2Storage(StorageBackend):
    """Cloudflare R2 storage for production (S3-compatible)."""

    def __init__(
        self,
        account_id: str,
        access_key: str,
        secret_key: str,
        bucket: str,
    ):
        self.bucket = bucket
        self.client = boto3.client(
            "s3",
            endpoint_url=f"https://{account_id}.r2.cloudflarestorage.com",
            aws_access_key_id=access_key,
            aws_secret_access_key=secret_key,
        )

    async def upload(self, path: str, data: bytes, content_type: str) -> str:
        self.client.put_object(
            Bucket=self.bucket,
            Key=path,
            Body=data,
            ContentType=content_type,
        )
        return path

    async def download(self, path: str) -> bytes:
        response = self.client.get_object(Bucket=self.bucket, Key=path)
        return response["Body"].read()

    async def delete(self, path: str) -> None:
        self.client.delete_object(Bucket=self.bucket, Key=path)

    def get_url(self, path: str, expires_in: int = 3600) -> str:
        return self.client.generate_presigned_url(
            "get_object",
            Params={"Bucket": self.bucket, "Key": path},
            ExpiresIn=expires_in,
        )

    async def exists(self, path: str) -> bool:
        try:
            self.client.head_object(Bucket=self.bucket, Key=path)
            return True
        except self.client.exceptions.ClientError:
            return False


def get_storage() -> StorageBackend:
    """Factory function to get storage backend based on environment."""
    backend = os.getenv("STORAGE_BACKEND", "local")

    if backend == "r2":
        return R2Storage(
            account_id=os.environ["R2_ACCOUNT_ID"],
            access_key=os.environ["R2_ACCESS_KEY_ID"],
            secret_key=os.environ["R2_SECRET_ACCESS_KEY"],
            bucket=os.environ["R2_BUCKET_NAME"],
        )

    return LocalStorage()

File: modules/agent-go/internal/storage/storage.go
package storage

import (
	"fmt"
	"io"
	"os"
	"path/filepath"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
)

// Backend defines the storage interface
type Backend interface {
	Upload(path string, data io.Reader, contentType string, size int64) error
	Download(path string) ([]byte, error)
	Delete(path string) error
	GetURL(path string) string
	Exists(path string) (bool, error)
}

// LocalStorage implements filesystem storage for local development
type LocalStorage struct {
	basePath string
}

// NewLocalStorage creates a new local storage backend
func NewLocalStorage(basePath string) *LocalStorage {
	if basePath == "" {
		basePath = os.Getenv("STORAGE_PATH")
		if basePath == "" {
			basePath = "/app/storage/documents"
		}
	}
	os.MkdirAll(basePath, 0755)
	return &LocalStorage{basePath: basePath}
}

func (s *LocalStorage) Upload(path string, data io.Reader, contentType string, size int64) error {
	fullPath := filepath.Join(s.basePath, path)
	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
		return err
	}

	file, err := os.Create(fullPath)
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = io.Copy(file, data)
	return err
}

func (s *LocalStorage) Download(path string) ([]byte, error) {
	return os.ReadFile(filepath.Join(s.basePath, path))
}

func (s *LocalStorage) Delete(path string) error {
	fullPath := filepath.Join(s.basePath, path)
	if _, err := os.Stat(fullPath); os.IsNotExist(err) {
		return nil
	}
	return os.Remove(fullPath)
}

func (s *LocalStorage) GetURL(path string) string {
	return fmt.Sprintf("file://%s", filepath.Join(s.basePath, path))
}

func (s *LocalStorage) Exists(path string) (bool, error) {
	_, err := os.Stat(filepath.Join(s.basePath, path))
	if os.IsNotExist(err) {
		return false, nil
	}
	return err == nil, err
}

// R2Storage implements Cloudflare R2 storage (S3-compatible)
type R2Storage struct {
	client *s3.S3
	bucket string
}

// NewR2Storage creates a new R2 storage backend
func NewR2Storage(accountID, accessKey, secretKey, bucket string) (*R2Storage, error) {
	endpoint := fmt.Sprintf("https://%s.r2.cloudflarestorage.com", accountID)

	sess, err := session.NewSession(&aws.Config{
		Endpoint:         aws.String(endpoint),
		Region:           aws.String("auto"),
		Credentials:      credentials.NewStaticCredentials(accessKey, secretKey, ""),
		S3ForcePathStyle: aws.Bool(true),
	})
	if err != nil {
		return nil, err
	}

	return &R2Storage{
		client: s3.New(sess),
		bucket: bucket,
	}, nil
}

func (s *R2Storage) Upload(path string, data io.Reader, contentType string, size int64) error {
	_, err := s.client.PutObject(&s3.PutObjectInput{
		Bucket:        aws.String(s.bucket),
		Key:           aws.String(path),
		Body:          aws.ReadSeekCloser(data),
		ContentType:   aws.String(contentType),
		ContentLength: aws.Int64(size),
	})
	return err
}

func (s *R2Storage) Download(path string) ([]byte, error) {
	resp, err := s.client.GetObject(&s3.GetObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(path),
	})
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return io.ReadAll(resp.Body)
}

func (s *R2Storage) Delete(path string) error {
	_, err := s.client.DeleteObject(&s3.DeleteObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(path),
	})
	return err
}

func (s *R2Storage) GetURL(path string) string {
	req, _ := s.client.GetObjectRequest(&s3.GetObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(path),
	})
	url, _ := req.Presign(3600)
	return url
}

func (s *R2Storage) Exists(path string) (bool, error) {
	_, err := s.client.HeadObject(&s3.HeadObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(path),
	})
	if err != nil {
		return false, nil
	}
	return true, nil
}

// GetStorage returns the appropriate storage backend based on environment
func GetStorage() Backend {
	backend := os.Getenv("STORAGE_BACKEND")

	if backend == "r2" {
		storage, err := NewR2Storage(
			os.Getenv("R2_ACCOUNT_ID"),
			os.Getenv("R2_ACCESS_KEY_ID"),
			os.Getenv("R2_SECRET_ACCESS_KEY"),
			os.Getenv("R2_BUCKET_NAME"),
		)
		if err != nil {
			panic(fmt.Sprintf("failed to initialize R2 storage: %v", err))
		}
		return storage
	}

	return NewLocalStorage("")
}
